### PR TITLE
More Solidity -> yul codegen

### PR DESCRIPTION
**Pull request by chriseth**
Part of #8343.

### CODE REVIEW

File: libsolidity/codegen/ir/IRGenerationContext.cpp
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGenerationContext.cpp]:
```suggestion
					functions.emplace_back(map<string, string> {
						{ "funID", to_string(function->id()) },
						{ "name", functionName(*function) }
					});
```
? Could even loose the braces (of the ``if``) then.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: This could be moved into the ``if`` case below, right? The ``else`` will be just fine for multiple declarations already.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:73]: Actually we should at least assert that the type fits in one variable, but since we don't have real type conversions anyways and would actually need that here, I guess having a TODO here is fine... generally this code is not yet expected to even properly assert everywhere, right?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Maybe ``std::find`` or something like that?

**Review by ekpyron — APPROVED** [review]
Only some minor comments and suggestions - but could also be merged as is.

Should we have tests for the generated Yul code (that e.g. test for structural equality) complementary to the semantics tests?

File: test/libsolidity/semanticTests/viaYul/detect_add_overflow.yul
Comment (ekpyron) [inline test/libsolidity/semanticTests/viaYul/detect_add_overflow.yul:11]: One of the failure cases here seem to succeed...

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: test/libsolidity/semanticTests/viaYul/detect_add_overflow.yul
Comment (ekpyron) [reply test/libsolidity/semanticTests/viaYul/detect_add_overflow.yul:11]: And I don't really see why...

Comment (ekpyron) [reply test/libsolidity/semanticTests/viaYul/detect_add_overflow.yul:11]: It's all of them in fact...

Comment (ekpyron) [reply test/libsolidity/semanticTests/viaYul/detect_add_overflow.yul:11]: Ah! Doh :-D! These don't have the ``compileViaYul`` moved down to the expectation blocks, so they're not testing this PR at all :-D.

Comment (ekpyron) [reply test/libsolidity/semanticTests/viaYul/detect_add_overflow.yul:11]: Just moved them down myself.

**Review by ekpyron — DISMISSED** [review]

File: libsolidity/codegen/YulUtilFunctions.h
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.h]:
```suggestion
	std::string overflowCheckedUIntAddFunction(size_t _bits);
```

File: libsolidity/codegen/ir/IRGenerationContext.cpp
Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: Sounds good!

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Yep, should use `std::find`.

**Review by ekpyron — DISMISSED** [review]
I'd be fine with this now, but since I edited it we should have another review.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:40]: Why was this moved inside? It doesn't depend on `expression`

**Comment by Marenz** [comment]
I fixed the compilation errors

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:40]: Because the ``else`` case works just fine for multi-variable declarations - only the declaration-and-assignment case is unimplemented for the multi-variable case.

**Comment by ekpyron** [comment]
Thanks @Marenz :-).
The last two commits should be squashed before this is merged.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:74]: Shouldn't this be a `solUnimplementedAssert`? I mean, in case the left hand side isn't an `Identifier`.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:75]: Same here

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I think this should also use `solUnimplementedAssert`.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:74]: I'm not sure - these dynamic casts will throw exceptions anyways, if they fail, and we have the ``TODO``s - do you think we should require this code to have proper assertions everywhere already?

Comment (leonardoalt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:74]: Ok, I guess it's fine without.

**Review by leonardoalt — DISMISSED** [review]

**Comment by leonardoalt** [comment]
Please remember to squash

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:74]: It's only here I think and this function will need a lot of work anyways, so yeah, I think it's fine.

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:74]: Ah, ok... the ``The function can very well return multiple vars.`` TODO's don't assert as well... still, I'd leave it at that for this PR.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (Marenz) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:43]: I believe none of the tests are actually testing this particular block here

Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:43]: I added some output here and ran all the tests and none of them triggered the output..

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:162]: Shouldn't this be called before the if above?

Comment (leonardoalt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:162]: Ah no, got it now

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:43]: That's true - you'd need a test case containing a local variable declaration with an assignment, e.g.
```
contract C {
    function f(uint a) public pure returns (uint x) {
        uint b = a;
        x = b;
    }
}
```
We should add that and see if it works.

Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:43]: Yep, I can confirm that this does trigger the block

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Please add tests that use `Bool`, `address`

**Review by leonardoalt — DISMISSED** [review]
Looks good. Please squash.

**Comment by Marenz** [comment]
squashed

**Comment by Marenz** [comment]
CI failure is a random one, re-triggering

**Review by ekpyron — APPROVED** [review]

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**
